pkg/ingester/ingester.go

```diff
@@ -502,6 +545,55 @@ func (i *Ingester) append(ctx context.Context, userID string, labels labelPairs,
 	return err
 }
 
+func (i *Ingester) appendMetadata(userID string, m *client.MetricMetadata) error {
+	i.userStatesMtx.RLock()
+	if i.stopped {
+		i.userStatesMtx.RUnlock()
+		return fmt.Errorf("ingester stopping")
+	}
+	i.userStatesMtx.RUnlock()
+
+	i.usersMetadataMtx.Lock()
+	userMetadata := i.getOrCreateUserMetadata(userID)
+	i.usersMetadataMtx.Unlock()
```

Comment:
This locking should be moved inside `getOrCreateUserMetadata()` so that we can implement it taking a read lock in most of the cases:

```
func (i *Ingester) getOrCreateUserMetadata(userID string) *userMetricsMetadata {
	i.usersMetadataMtx.RLock()
	userMetadata, ok := i.usersMetadata[userID]
	i.usersMetadataMtx.RUnlock()
	
	if ok {
		return userMetadata
	}
	
	i.usersMetadataMtx.Lock()
	defer i.usersMetadataMtx.Unlock()
	
	// Ensure it has not been created in the meanwhile.
	userMetadata, ok = i.usersMetadata[userID]
	if !ok {
		userMetadata = newMetadataMap(i.limiter, i.metrics, userID)
		i.usersMetadata[userID] = userMetadata
	}

	return userMetadata
}
```